### DISCUSSION TITLE
Collection record response is not serializable

### DISCUSSION THREAD

**Discussion by LukasParke** [pocketbase/pocketbase#1372] [Q&A]
My example use case here is Sveltekit server layout load functions.

```js
const pb = new PocketBase(pocketUrl);

// fetch a paginated records list
let posts: Record[] = [];

posts = (
	await pb.collection('posts').getList(1, 20, {
		sort: '-created'
	})
).items;

console.log(posts);
```

This snippet returns a record structure for me of
```json
n {
  body: 'Test Article Body',
  collectionId: 'ae6u3fqy720uas6',
  collectionName: 'posts',
  created: '2022-12-24 15:22:01.285Z',
  id: 'g3e5bhwxlv91rer',
  live: true,
  photos: [ '1738_3_XtsV3aJ1Jq.png' ],
  title: 'Test Article Title',
  updated: '2022-12-24 15:22:01.285Z',
  expand: {}
}
```

which returning in sveltekit throws this error:
```shell
Data returned from `load` while rendering null is not serializable: Cannot stringify arbitrary non-POJOs (data.posts[0])
```

My current working solution to mitigate this error is to manually stringify and parse the object before returning from the load function, below is my full implementation for this example
```js
export const load = (async () => {
const pb = new PocketBase(pocketUrl);

// fetch a paginated records list
let posts: Record[] = [];

posts = (
	await pb.collection('posts').getList(1, 20, {
		sort: '-created'
	})
).items;

console.log(posts);

posts = posts.map((obj) => {
	console.log(obj);
	console.log(JSON.stringify(obj));

	return JSON.parse(JSON.stringify(obj));
});

console.log(posts);

return { posts };

}) satisfies LayoutLoad;
```

after sanitization, this is the structure of the record array:
```json
[
  {
    body: 'Test Article Body',
    collectionId: 'ae6u3fqy720uas6',
    collectionName: 'posts',
    created: '2022-12-24 15:22:01.285Z',
    id: 'g3e5bhwxlv91rer',
    live: true,
    photos: [ '1738_3_XtsV3aJ1Jq.png' ],
    title: 'Test Article Title',
    updated: '2022-12-24 15:22:01.285Z',
    expand: {}
  }
]
```

**Comment by ganigeorgiev** [comment]
This is expected. You can check https://github.com/pocketbase/pocketbase/discussions/1025.

`getList()` returns a [`ListResult` class instance](https://github.com/pocketbase/js-sdk/blob/master/src/models/utils/ListResult.ts) not a plain object. The `items` inside the list is also an array of `Record` objects.

As the error message indicates you need a seriazable data and your workaround is correct.
Eventually you can replace `JSON.parse(JSON.stringify(obj));` with `structuredClone(obj)` (if you are targetting modern browsers and/or using node17+; also instead of `obj` you can pass directly the entire `posts` array if you returning it anyway).

**Comment by LukasParke** [comment]
Thanks for the response Gani, I would normally expect a JavaScript sdk to
return standard JavaScript object notation, but I appreciate your
clarification that it’s expected.

**Reply by ganigeorgiev** [reply]
@LukeHagar I don't understand. The SDK is returning a JS class, it just not a POJO. It is returning the type you are using in your TS `posts` declaration, aka. `Record[]`.

The error you are getting is just an implementation detail of the `load` method. It wants:

```js
[{...}, {...}, {...}, ...]
```

but instead we are returning:

```js
[new Record(), new Record(), new Record(), ...]
```

There is nothing special or "magic" here, it is a standard JS object. If you need  a POJO, just wrap the return result with `structuredClone(posts)`.

**Reply by ghost** [reply]
Hey Gani,

I understand what you are saying, but the api calls return standard JSON don't they?
```json
{"page":1,"perPage":30,"totalItems":1,"totalPages":1,"items":[{"body":"Test Article Body","collectionId":"ae6u3fqy720uas6","collectionName":"posts","created":"2022-12-24 15:22:01.285Z","id":"g3e5bhwxlv91rer","live":true,"photos":["1738_3_XtsV3aJ1Jq.png"],"title":"Test Article Title","updated":"2022-12-24 15:22:01.285Z"}]}
```
I just don't understand why you would convert it to a class before returning from the function?

I am certain there is some benefit to it I'm not understanding, and I apologize for my ignorance, I'm just confused as to why its returning what it is.

regardless the workarounds discussed are functional for now, and I am loving pocketbase as a whole, thank you for working on something so cool!

**Reply by ganigeorgiev** [reply]
> I just don't understand why you would convert it to a class before returning from the function?

Because there are various helper methods attached to it that also take care for normalizing the api response (or any other custom user defined data).

For example, you can explore the `load()` of `BaseModel` and `Record` class methods (used in the constructor):

- https://github.com/pocketbase/js-sdk/blob/master/src/models/utils/BaseModel.ts
- https://github.com/pocketbase/js-sdk/blob/master/src/models/Record.ts

Having a dedicated class also helps with applying api changes in a backward compatible manner more easily (eg. https://github.com/pocketbase/js-sdk/blob/master/src/models/LogRequest.ts).

**Reply by ghost** [reply]
Thank you for explaining Gani, seems like a small price to pay for more control over normalization, I think i'll just make a helper function to wrap all my returns in, but I am extremely excited to dive further into pocketbase.

**Reply by ganigeorgiev** [reply]
Please note that you don't always need to wrap the result. It is just a specific of SvelteKit probably so that they can "move" the data in a more unified way without worrying that some custom implementation may not be available in the different context where the data could be used (eg. when moving data from node to the browser or to a worker). Per https://kit.svelte.dev/docs/load#universal-vs-server-output:

> A server load function must return data that can be serialized with [devalue](https://github.com/rich-harris/devalue) — anything that can be represented as JSON plus things like BigInt, Date, Map, Set and RegExp, or repeated/cyclical references — so that it can be transported over the network.

Also note that there is a "universal" `load` function (aka. the POJO requirement is only for the server load):

> A universal load function can return an object containing **any values**, including things like **custom classes** and component constructors.

---

Outside of these specific cases, for the end user it shouldn't matter too much what is the type of the object, since accessing the fields is the same as when working with POJO, eg. `record.id`.
